primitives/io/src/async_externalities.rs

```diff
@@ -0,0 +1,217 @@
+// This file is part of Substrate.
+
+// Copyright (C) 2020 Parity Technologies (UK) Ltd.
+// SPDX-License-Identifier: GPL-3.0-or-later WITH Classpath-exception-2.0
+
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see <https://www.gnu.org/licenses/>.
+
+//! Async externalities.
+
+use std::any::{TypeId, Any};
+use sp_core::{
+	storage::{ChildInfo, TrackedStorageKey},
+	traits::{Externalities, SpawnNamed, TaskExecutorExt},
+};
+use crate::{RuntimeSpawnExt, RuntimeSpawn};
+use sp_externalities::{Extensions, ExternalitiesExt as _};
+
+/// Simple state-less externalities for use in async context.
+///
+/// Will panic if anything is accessing the storage.
+#[derive(Debug)]
+pub struct AsyncExternalities {
+	extensions: Extensions,
+}
+
+/// New Async externalities.
+pub fn new_async_externalities(scheduler: Box<dyn SpawnNamed>) -> Result<AsyncExternalities, &'static str> {
+	let extensions = Extensions::default();
+	let mut res = AsyncExternalities { extensions };
+	let mut ext = &mut res as &mut dyn Externalities;
+	ext.register_extension::<TaskExecutorExt>(TaskExecutorExt(scheduler.clone()))
+		.map_err(|_| "Failed to register task executor extension.")?;
+
+	Ok(res)
+}
+
+impl AsyncExternalities {
+	/// Extend async externalities with the ability to spawn wasm instances.
+	pub fn with_runtime_ext(
+		mut self,
+		runtime_ext: Box<dyn RuntimeSpawn>,
+	) -> Result<Self, &'static str> {
+		let mut ext = &mut self as &mut dyn Externalities;
+		ext.register_extension::<RuntimeSpawnExt>(RuntimeSpawnExt(runtime_ext))
+			.map_err(|_| "Failed to register task executor extension.")?;
+
+		Ok(self)
+	}
+}
+
+type StorageKey = Vec<u8>;
+
+type StorageValue = Vec<u8>;
```

Comment:
I would put these and next to the definition of `trait Externalities` and have all impls use them, but well, anyways they're all gonna be Vec<u8> probably for ever anyhow.